include/xtensor/xmath.hpp

```diff
@@ -3013,6 +3013,86 @@ namespace detail {
             return cov(eval(stack(xtuple(x, y))));
         }
     }
+
+    /*
+    * convolution mode placeholders
+    */
+    namespace conv_mode
+    {
+        struct valid{};
+        struct full {};
+    }
+
+    /*
+    * @brief computes the 1D convolution between two 1D expressions
+    *
+    * @param a 1D expression
+    * @param v 1D convolutional kernel
+    * @param mode placeholder from conv_mode namespace
+    * 
+    * @detail the algorithm convolves a with v and will incur a copy overhead
+    *   should v be longer than a.
+    */
+    template <class E1, class E2, class E3>
+    inline auto convolve(E1&& a, E2&& v, E3&& mode)
+    {
+        using value_type = typename std::decay<E1>::type::value_type;
+
+        if (a.shape().size() != 1 || v.shape().size() != 1)
+        {
+            XTENSOR_THROW(std::runtime_error, "Invalid dimentions convolution arguments must be 1D expressions");
+        }
+
+        if (a.shape(0) == 0 || v.shape(0) == 0)
+        {
+            XTENSOR_THROW(std::runtime_error, "Arguments cannot have zero length");
+        }
+
+        //swap them so a is always the longest one
+        if (a.shape(0) < v.shape(0))
+        {
+            auto temp = a;
+            a = v;
+            v = temp;
+        }
```

Comment:
A mean to avoid copy is to have an additional function performing the implementation that is called by this one:

```cpp
namespace detail
{
    template <class E1, class E2, class E3>
    inline auto convolve_impl(E1&& e1, E2&& e2, E3 mode)
    {
        // Implementation with no checks
    }
}

template <class E1, class E2, class E3>
inline auto convolve(E1&& a, E2&& b, E3 mode)
{
    // Dimensions and shape checks
    // ...

    bool cond = a.shape(0) < v.shape(0);
    return cond ? detail::convole_impl(std::forward<E2>(v), std::forward<E1>(a), mode)
                       : detail::convole_impl(std::forward<E1>(a), std::forward<E2>(v), mode)
}